nasl/nasl_crypto2.c

```diff
@@ -1493,6 +1493,157 @@ nasl_bf_cbc_decrypt (lex_ctxt *lexic)
   return nasl_bf_cbc (lexic, 0);
 }
 
+/**
+ * @brief Open a stream cipher. This function creates a context handle and
+ * stores it in a cipher table.
+ * Open cipher must be deleted with delete_cipher_item() at the end of the
+ * stream encryption.
+ * @param[in] cipher The cipher algorithm.
+ * @param[in] mode The cipher mode. Must be compatible with the algorithm.
+ * @return Returns the ID of the cipher handler on success. Otherwise NULL.
+ */
+static tree_cell *
+nasl_open_stream_cipher (lex_ctxt *lexic, int cipher, int mode)
+{
+  gcry_cipher_hd_t hd;
+  gcry_error_t error;
+  void *key, *iv;
+  size_t keylen, ivlen;
+  tree_cell *retc;
+  static int cipher_id = 32;
+  cipher_table_item_t *hd_item;
+
+  key = get_str_var_by_name (lexic, "key");
+  keylen = get_var_size_by_name (lexic, "key");
+  iv = get_str_var_by_name (lexic, "iv");
+  ivlen = get_var_size_by_name (lexic, "iv");
+
+  if (!key || keylen <= 0)
+    {
+      nasl_perror (lexic, "Syntax: encrypt_data: Missing data or key argument");
+      return NULL;
+    }
+
+  if ((error = gcry_cipher_open (&hd, cipher, mode, 0)))
+    {
+      nasl_perror (lexic, "gcry_cipher_open: %s", gcry_strerror (error));
+      gcry_cipher_close (hd);
+      return NULL;
+    }
+  if ((error = gcry_cipher_setkey (hd, key, keylen)))
+    {
+      nasl_perror (lexic, "gcry_cipher_setkey: %s", gcry_strerror (error));
+      gcry_cipher_close (hd);
+      return NULL;
+    }
+
+  if (iv && ivlen)
+    {
+      if ((error = gcry_cipher_setiv (hd, iv, ivlen)))
+        {
+          nasl_perror (lexic, "gcry_cipher_setiv: %s", gcry_strerror (error));
+          return NULL;
+        }
+    }
+
+  hd_item = cipher_table_item_new ();
+  hd_item->hd = hd;
+  hd_item->id = cipher_id;
+  cipher_id--;
```

Comment:
I'm a bit unsure about this.  nasl_close_stream_cipher does not increment the value, so could the NVT run out of ids?  And there's no check that cipher_id has reached 0.